src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncDisposableTests.cs

```diff
@@ -0,0 +1,900 @@
+    IL_008f:  leave      IL_0125
+    IL_0094:  ldarg.0
+    IL_0095:  ldfld      ""System.Runtime.CompilerServices.TaskAwaiter S.<Main>d__0.<>u__1""
+    IL_009a:  stloc.2
+    IL_009b:  ldarg.0
+    IL_009c:  ldflda     ""System.Runtime.CompilerServices.TaskAwaiter S.<Main>d__0.<>u__1""
+    IL_00a1:  initobj    ""System.Runtime.CompilerServices.TaskAwaiter""
+    IL_00a7:  ldarg.0
+    IL_00a8:  ldc.i4.m1
+    IL_00a9:  dup
+    IL_00aa:  stloc.0
+    IL_00ab:  stfld      ""int S.<Main>d__0.<>1__state""
+    IL_00b0:  ldloca.s   V_2
+    IL_00b2:  call       ""void System.Runtime.CompilerServices.TaskAwaiter.GetResult()""
+    IL_00b7:  nop
+    IL_00b8:  ldarg.0
+    IL_00b9:  ldfld      ""object S.<Main>d__0.<>s__2""
+    IL_00be:  stloc.1
+    IL_00bf:  ldloc.1
+    IL_00c0:  brfalse.s  IL_00dd
+    IL_00c2:  ldloc.1
+    IL_00c3:  isinst     ""System.Exception""
+    IL_00c8:  stloc.s    V_4
+    IL_00ca:  ldloc.s    V_4
+    IL_00cc:  brtrue.s   IL_00d0
+    IL_00ce:  ldloc.1
+    IL_00cf:  throw
+    IL_00d0:  ldloc.s    V_4
+    IL_00d2:  call       ""System.Runtime.ExceptionServices.ExceptionDispatchInfo System.Runtime.ExceptionServices.ExceptionDispatchInfo.Capture(System.Exception)""
+    IL_00d7:  callvirt   ""void System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()""
+    IL_00dc:  nop
+    IL_00dd:  ldarg.0
+    IL_00de:  ldfld      ""int S.<Main>d__0.<>s__3""
+    IL_00e3:  stloc.s    V_5
+    IL_00e5:  ldloc.s    V_5
+    IL_00e7:  ldc.i4.1
+    IL_00e8:  beq.s      IL_00ec
+    IL_00ea:  br.s       IL_00ee
+    IL_00ec:  leave.s    IL_0111
+    IL_00ee:  ldarg.0
+    IL_00ef:  ldnull
+    IL_00f0:  stfld      ""object S.<Main>d__0.<>s__2""
+    IL_00f5:  leave.s    IL_0111
+  }
+  catch System.Exception
+  {
+    IL_00f7:  stloc.s    V_4
+    IL_00f9:  ldarg.0
+    IL_00fa:  ldc.i4.s   -2
+    IL_00fc:  stfld      ""int S.<Main>d__0.<>1__state""
+    IL_0101:  ldarg.0
+    IL_0102:  ldflda     ""System.Runtime.CompilerServices.AsyncTaskMethodBuilder S.<Main>d__0.<>t__builder""
+    IL_0107:  ldloc.s    V_4
+    IL_0109:  call       ""void System.Runtime.CompilerServices.AsyncTaskMethodBuilder.SetException(System.Exception)""
+    IL_010e:  nop
+    IL_010f:  leave.s    IL_0125
+  }
+  IL_0111:  ldarg.0
+  IL_0112:  ldc.i4.s   -2
+  IL_0114:  stfld      ""int S.<Main>d__0.<>1__state""
+  IL_0119:  ldarg.0
+  IL_011a:  ldflda     ""System.Runtime.CompilerServices.AsyncTaskMethodBuilder S.<Main>d__0.<>t__builder""
+  IL_011f:  call       ""void System.Runtime.CompilerServices.AsyncTaskMethodBuilder.SetResult()""
+  IL_0124:  nop
+  IL_0125:  ret
+}");
+        }
+
+        [Fact]
+        public void TestWithNullableExpression()
+        {
+            string source = @"
+struct S : System.IAsyncDisposable
+{
+    public static async System.Threading.Tasks.Task Main()
+    {
+        S? s = new S();
+        using await (s)
+        {
+            System.Console.Write(""body "");
+            return;
+        }
```

Comment:
Consider testing `S? s = null; using await (s) { ... }`.

Reply:
Thanks for the test suggestions (good ideas). Added them

---
In reply to: [160596740](https://github.com/dotnet/roslyn/pull/23961#discussion_r160596740) [](ancestors = 160596740)